drivers: add driver for ST VL6180X ranging and ambient light sensor

gschorcht opened the pull request:
### Contribution description

This PR adds a driver for the ST VL6180X ranging and ambient light sensor. The driver can be used in two variants which differ in functionality and size:

Module Name   | Driver
:-------------|:-------------------------------------------
vl6180x       | Driver with standard functionality and larger size
vl6180x_basic | Driver with only basic functionality and smaller size

Both driver variants provide

- continuous ranging and ambient light sensing,
- polling for new measurement results, and
- SAUL capabilies.

In addition to these basic functionalities, the standard driver `vl6180x`
 provides interrupt handling, single shot measurements, a number of
 configuration functions and power down / power up functionality.
 Features of the driver variants are summarized in the following comparison
 sheet. 

Feature                                 | vl6180x_basic | vl6180x
:---------------------------------------|:-------------:|:--------:
Range measurements [mm]                 | X             | X
Ambient light measurements [Lux, raw]   | X             | X
SAUL capability                         | X             | X
Continuous measurements                 | X             | X
Single shot measurements                |               | X
Start and stop measurements             |               | X
Polling for new measurement results     | X             | X
Data-ready interrupts                   |               | X
Threshold interrupts                    |               | X
Measurement error status                |               | X
Configuration during run-time           |               | X
Power down and power up                 |               | X
Size on reference platform              | 1.7 kByte     | 3.0 kByte

### Testing procedure

Using the make command
```
make flash -C tests/driver_vl6180x BOARD=...
```
the standard driver variant `vl6180x` is used with the following default
configuration parameters:

- periodic check of the availability of new data every 100 ms
- a ranging inter-measurement period of 100 ms
- a ranging maximum convergence time of 50 ms
- an ambient light sensing (ALS) inter-measurement period of 500 ms
- an ambient light sensing (ALS) integration time of 100 ms
- an ambient light sensing (ALS) analogue light channel gain of 1.0
- an ambient light sensing period of 500 ms
- a ranging period of 100 ms.

To use the basic driver variant, the `vl6180x_basic` module has
to be specified at make command line
```
USEMODULE=vl6180x_basic make flash -C tests/driver_vl6180x BOARD=...
```
The default configuration parameters used for the test application with the
basic driver variant are the same as for the standard driver variant.

To use interrupts to fetch new data instead of polling for new data
periodically, the standard driver variant `vl6180x` has to be used and the
pin connected with the interrupt signal GPIO1 of the sensor has to be
defined by configuration paramater `VL6180X_PARAM_PIN_INT`, e.g.,
```
CFLAGS="-DVL6180X_PARAM_PIN_INT=\(GPIO_PIN\(0,1\)\)" make flash -C tests/driver_vl6180x BOARD=...
```
To test the power down and power up functionality, the pin connected with
the signal GPIO0/CE of the sensor has to be defined by configuration
paramater `VL6180X_PARAM_PIN_SHUTDOWN
`, e.g.,
```
CFLAGS="-DVL6180X_PARAM_PIN_SHUTDOWN=\(GPIO_PIN\(0,2\)\)" make flash -C tests/driver_vl6180x BOARD=...
```

aabadie commented:
I have the hardware for testing this PR, I'll do a code review this week.

gschorcht commented:
@aabadie Thanks :smile:

gschorcht commented on drivers/include/vl6180x.h:
```diff
@@ -0,0 +1,644 @@
+} vl6180x_error_t;
+
+/**
+ * @brief   Aanalogue gain of ALS measurements
```
Thanks

gschorcht commented on drivers/include/vl6180x.h:
```diff
@@ -0,0 +1,644 @@
+ */
+typedef enum {
+    VL6180X_ALS_GAIN_20 = 0,    /**<   20 x gain */
+    VL6180X_ALS_GAIN_100,       /**<   10 x gain */
```
Yes, of course :dark_sunglasses:

gschorcht commented on tests/driver_vl6180x/main.c:
```diff
@@ -0,0 +1,260 @@
+    uint16_t als;
+    uint16_t lux;
+
+#if MODULE_VL6180X_BASIC
```
Why, the semantic is the same? By definition, an undefined macro can be tested for a zero. That is `#if !` is exactly the same as `#ifndef` and `#if !defined()`.

dylad commented:
@gschorcht Could you please rebase this PR ?
I just received the hardware to test it so I can give it a try.

gschorcht commented:
> @gschorcht Could you please rebase this PR ?
> I just received the hardware to test it so I can give it a try.

I'm on vacation until August 27th. Is it ok to rebase it afterwards?

dylad commented:
No problem ! Enjoy your vacations !

gschorcht commented:
@leandrolanzieri As part of a complete rework of this driver I will also provide the `Kconfig` file. Is there any possibility to define a float/double configuration parameter? How would you realize it. My first idea was to define it as integer, for example multiplied by factor 1000.

benpicco commented:
If float can be avoided altogether that would be idea - soft float can be rather expensive on low-end devices.

gschorcht commented:
> If float can be avoided altogether that would be ideal - soft float can be rather expensive on low-end devices.

That is clear. But when raw sensor data must be converted from something that means nothing to something real, sometimes the only way is to use floating point or at least fixed point arithmetic. An example is the conversion from counts (which mean nothing) to lux. The conversion depends on different factors that are mostly given as floats like the gain chosen or the lux resolution which in turn depend on the cover glass used.

leandrolanzieri commented:
> Is there any possibility to define a float/double configuration parameter? How would you realize it. My first idea was to define it as integer, for example multiplied by factor 1000.

Kconfig does not allow for a float parameter type, so I would go with the dividing factor. I did not look at the specifics of the driver, but if you want the dividing factor also configurable, you can limit its values by a `choice`, and based on the option chosen apply a valid `range <> <> if <>` to the `int` parameter.

gschorcht commented:
> If float can be avoided altogether that would be ideal - soft float can be rather expensive on low-end devices.

I think the solution is to use integers, which are simply the float values multiplied by 1000. Since lux values are computed by
```
lux = lux_res * (cnt / als_gain) * (100 / als_int_time)
```
where `lux_res` and `als_gain` are floats, using these values multiplied by 1000 should work. The resolution should also be sufficient, since both floats are specified with a resolution of only 1/100. However, because of the value range of the calculation `uint32_t` is required, which is probably still better than `float`.

gschorcht commented:
@aabadie @dylad @smlng  After the complete rework is finished, almost everything has changed. So from my point of view it makes no sense to provide the changes as fixups to the original PR. Also, because `Kconfig` files were added, I would restructure the commits a bit. Would it be ok to force-push the squashed and restructured commits?

dylad commented:
@gschorcht 
No problem, go ahead !

dylad commented:
@gschorcht I have this module on my desk wired to a SAM board and ready to be tested.
If you're still willing to push it forward. Feels free to rebase/squash.

gschorcht commented:
@dylad The driver is now ready for the next round of testing. I rebased and squashed it. I had to fix a few new compilation problems that came up after rebasing and squashed them directly.

dylad commented:
First testing session:
Ranging seems to work pretty good in polling and interrupt mode.
But ALS is stuck at 0 for both mode. Am I missing something ?

dylad commented:
Note to myself: get off your cave.

gschorcht commented:
Strange. Even in a pretty dark environment I get something like the following in both modes. I just connect the sensor breakout board to the MCU board, that's it.:
```
main(): This is RIOT! (Version: 2023.01-devel-648-g00a8f4-drivers_vl6180x)
VL6180X Time-of-Flight distance sensor

Initializing VL6180X sensor
[OK]
ALS: 51 [cnts], 16 [lux]
ALS: 51 [cnts], 16 [lux]
RNG: 33 [mm]
+-----------------------------------------+
ALS: 51 [cnts], 16 [lux]
RNG: 33 [mm]
+-----------------------------------------+
ALS: 51 [cnts], 16 [lux]
...
```
I tested it for an ESP32 and a STM32F767ZI board. I will test it later today on an Arduino MKR1000 board.

dylad commented:
> Strange. Even in a pretty dark environment I get something like the following in both modes.

I'll give it another try this weekend. it's supposed to be sunny outside :sunny: 
I'm also preparing the first round of review :wink:

gschorcht commented:
For me, it works also with the Arduino MKR1000 board.

What seems a bit strange in your case is that the ALS measurements are performed even if the results are 0. Otherwise you would not get any ALS values at all. In both polling and interrupt modes, the corresponding results are read from the sensor and printed by the test application only when the sensor indicates that there is data to read.

dylad commented on drivers/vl6180x/Kconfig:
```diff
@@ -0,0 +1,213 @@
+
+if MODULE_VL6180X
+
+config MODULE_VL6180X_RNG
```
I'm a bit puzzled by the RNG abbreviation. I find it a bit disturbing at first glance.
I am not asking for a change here but I am wondering what other think about it.

gschorcht replied:
Ah ok, could be misinterpreted as random number generator. I din't think about it, I was just looking for short wording.

dylad commented on drivers/vl6180x/Kconfig:
```diff
@@ -0,0 +1,213 @@
+config VL6180X_RNG_INT_DRDY
+    bool "Data Ready"
+    help
+        Interrupt is triggered when new raning data are ready to be read
```
```suggestion
        Interrupt is triggered when new ranging data are ready to be read
```

gschorcht replied:
Hm, it didn't be catched by spellcheck, strange.

dylad commented on tests/driver_vl6180x/main.c:
```diff
@@ -0,0 +1,264 @@
+    puts("Initializing VL6180X sensor");
+
+    if (vl6180x_init(&dev, &vl6180x_params[0]) == VL6180X_OK) {
+        printf("[OK]\n");
```
```suggestion
        puts("[OK]");
```

gschorcht replied:
Optimization replaces `printf("...\n")` by `puts("...")` anyway.

dylad commented on tests/driver_vl6180x/main.c:
```diff
@@ -0,0 +1,264 @@
+        printf("[OK]\n");
+    }
+    else {
+        printf("[Failed]\n");
```
```suggestion
        puts("[Failed]");
```

dylad replied:
Same comments for others printf() below in this file.

dylad commented on tests/driver_vl6180x/main.c:
```diff
@@ -0,0 +1,264 @@
+
+#else /* IS_USED(MODULE_VL6180X_IRQ) */
+
+        /* just wait 200 ms if module vl6180x_basic is used */
+        ztimer_sleep(ZTIMER_MSEC, 250);
```
200 or 250 ?

gschorcht replied:
Let's use 250 ms.

dylad reviewed (commented):
First round of reviews. Mostly nitpicks.
Overall is really good.

gschorcht commented on drivers/vl6180x/Kconfig:
```diff
@@ -0,0 +1,213 @@
+    depends on TEST_KCONFIG
+    depends on HAS_PERIPH_I2C
+    select MODULE_PERIPH_I2C
+    select MODULE_ZTIMER
+    select MODULE_ZTIMER_MSEC
```
It is aslo used in `_init` function https://github.com/RIOT-OS/RIOT/blob/00a8f4d3ab0fd71ae465f5ce526e5eb12eb923ab/drivers/vl6180x/vl6180x.c#L653 to add the delay after power up as required according to the datasheet.

gschorcht commented on drivers/vl6180x/Kconfig:
```diff
@@ -0,0 +1,213 @@
+config VL6180X_RNG_THRESH_LOW
+    int "Ranging lower threshold [mm]"
+    depends on VL6180X_RNG_INT_LOW || VL6180X_RNG_INT_OUT
+    range 0 255
```
I'm not sure. The sensor allows to set the thresholds to 255 even if it doesn't make sense. But if you think it makes more sense to limit the values at user level, I can change it to 100.

dylad replied:
If the hardware allows I think we can stay as it then.

gschorcht commented on drivers/vl6180x/include/vl6180x_regs.h:
```diff
@@ -0,0 +1,152 @@
+#define VL6180X_CLR_ALL_INT     (0x07) /**< Clear all interrupts */
+
+/* VL6180X_REG_GPIO0_MODE */
+#define VL6180X_GPIO0_SHUT      (0x40) /**< GPIO1 shutdown function mask */
```
Copy & Paste error :sunglasses:

gschorcht commented on drivers/vl6180x/vl6180x.c:
```diff
@@ -0,0 +1,922 @@
+
+#endif /* ENABLE_DEBUG */
+
+#define ERROR_DEV(f, d, ...) \
+        LOG_ERROR("[vl6180x] %s i2c dev=%d addr=%02x: " f "\n", \
+                  __func__, d->params.i2c_dev, d->params.i2c_addr, ## __VA_ARGS__)
```
Not really clear what you mean.

dylad replied:
I was wondering why `ERROR_DEV` was using the log module while `DEBUG_DEV` was not.
While we are at it, you could use the log module for the whole file (including the debug level).

gschorcht replied:
If I'm right, the logic behind `DEBUG` together with `ENABLE_DEBUG` and `LOG_DEBUG` is different. `DEBUG` together with `ENABLE_DEBUG` only enables the debug output for the module while  `LOG_DEBUG` requires that `LOG_LEVEL` is greater than or equal 4 which enables the debug output for the whole application which isn't the intention in this case.

dylad commented:
Please squash !
Is `vl6180x_basic` pseudomodule a leftover from this PR description ?
Otherwise works as advertise.

gschorcht commented:
> Is `vl6180x_basic` pseudomodule a leftover from this PR description ?

Initially the driver had two versions selected by `vl6180x` and `vl6180x_basic`. While `vl6180x_basic` provided only basic functionally, `vl6180x` was fully featured. After the complete rework last year, `vl6180x` now provides the standard functionality and additional features can be enable by modules `vl6180x_irq`, `vl6180x_shutdown` and `vl6180x_config`. Thus, the PR description differes from the documentation of the current implementation.

Should I update the description?

dylad commented:
> Should I update the description?

I don't think so. The documentation is well written and this is the place we are supposed to look at anyways.
I'm pretty happy with the current state of this PR !

dylad reviewed (approved):
ACK.
I've tested this PR with a `VL8160X` module attached to a `SAML21-XPRO`
Works well, ranging measure seems accurate.

dylad commented:
bors merge

maribu commented:
bors retry

maribu commented:
bors retry

maribu commented:
bors cancel
bors retry

maribu commented:
@gschorcht Maybe a rebase can convince bors to pick this up?

benpicco commented:
bors merge

benpicco commented:
bors merge

benpicco commented:
This needs a rebase for bors to accept it.

benpicco commented:
bors merge

gschorcht commented:
Thanks :smile:
